infrastructure/src/main/java/org/wso2/testgrid/infrastructure/providers/AWSProvider.java

```diff
@@ -270,6 +290,81 @@ private InfrastructureProvisionResult doProvision(Script script, Properties inpu
         }
     }
 
+    /**
+     * Constructs the URL to Kibana dashboard to view logs.
+     *
+     * Makes use of the static dashboard having logs of all test plans. Filters to view logs
+     * of each EC2 instance are derived using placeholders. The default view shows all logs. The shortened URL of the
+     * dashboard is retrieved through Kibana ShortenURL API to have a compact url.
+     *
+     * @param testPlan test-plan to get log url for
+     * @param stackName name of the stack created for the test-plan
+     */
+    private void deriveLogDashboardUrl(TestPlan testPlan, String stackName) {
+        String kibanaEndpoint = ConfigurationContext.getProperty(ConfigurationProperties.KIBANA_ENDPOINT_URL);
+        String dashboardCtxFormat = TestGridConstants.KIBANA_DASHBOARD_STR;
+        String instanceLogFilterFormat = TestGridConstants.KIBANA_FILTER_STR;
+        String instanceLogFilter;
+        String instanceName;
+        StringJoiner filtersStr = new StringJoiner(",");
+
+        // Filter the EC2 instance corresponding to the stack
+        AmazonEC2 amazonEC2 = AmazonEC2ClientBuilder.defaultClient();
+        List<Reservation> reservations = amazonEC2.describeInstances().getReservations()
+                .stream().filter(r -> r.getInstances().get(0).getTags().contains(
+                        new Tag(STACK_NAME_TAG_KEY, stackName))).collect(Collectors.toList());
+        Map<String, String> instancesMap = new HashMap<>();
+
+        /* Create a map with instance ids and names to create the log filter. InstanceId is used to filter the index
+           pattern and name is used to set a label for the filter */
+        for (Reservation reservation : reservations) {
+            instanceName = reservation.getInstances().get(0).getTags().stream().filter(
+                    t -> t.getKey().equalsIgnoreCase(INSTANCE_NAME_TAG_KEY))
+                    .collect(Collectors.toList()).get(0).getValue();
+            instancesMap.put(reservation.getInstances().get(0).getInstanceId(), instanceName);
+        }
```

Comment:
We could also get for-loop as part of this stream. You'll need to use Stream#flatMap().